File: app/src/main/java/com/codepath/shapes/io/ShapeSerializer.java
package com.codepath.shapes.io;

import android.content.Context;
import android.os.AsyncTask;
import android.support.annotation.MainThread;
import android.support.annotation.NonNull;
import android.support.annotation.WorkerThread;
import android.util.Log;

import com.codepath.shapes.shape.Circle;
import com.codepath.shapes.shape.Rectangle;
import com.codepath.shapes.shape.Shape;
import com.codepath.shapes.shape.Text;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class ShapeSerializer {
    private static final String TAG = "ShapeSerializer";

    private static final String SHAPE_X = "x";
    private static final String SHAPE_Y = "y";
    private static final String SHAPE_COLOR = "color";
    private static final String SHAPE_TYPE = "type";
    private static final String CIRCLE_RADIUS = "radius";
    private static final String RECT_WIDTH = "width";
    private static final String RECT_HEIGHT = "height";
    private static final String TEXT_TEXT = "text";
    private static final String TEXT_FONT_SIZE = "fontSize";
    private static final String TYPE_CIRCLE = "circle";
    private static final String TYPE_RECTANGLE = "rectangle";
    private static final String TYPE_TEXT = "text";

    private static final String FILENAME = "shapes.json";


    @NonNull
    private Context mContext;


    public ShapeSerializer(@NonNull Context context) {
        mContext = context;
    }

    @MainThread
    public void saveShapes(@NonNull List<Shape> shapeList, @NonNull final OnSaveFinishedCallback callback) {
        final AsyncTask<List<Shape>, Void, Boolean> saveTask = new AsyncTask<List<Shape>, Void, Boolean>() {

            @SafeVarargs
            @Override
            protected final Boolean doInBackground(List<Shape>... params) {
                final List<Shape> shapeList = params[0];
                internalSaveShapes(shapeList);
                return true;
            }

            @Override
            protected void onPostExecute(Boolean aBoolean) {
                callback.onSaveFinished();
            }
        };

        saveTask.execute(shapeList);
    }


    @MainThread
    public void restoreShapes(@NonNull final OnRestoreFinishedCallback callback) {
        final AsyncTask<Void, Void, List<Shape>> restoreTask = new AsyncTask<Void, Void, List<Shape>>() {

            @Override
            protected List<Shape> doInBackground(Void... params) {
                return internalRestoreShapes();
            }

            @Override
            protected void onPostExecute(List<Shape> shapeList) {
                callback.onRestoreFinished(shapeList);
            }
        };

        restoreTask.execute();
    }

    @WorkerThread
    private void internalSaveShapes(@NonNull List<Shape> shapeList) {
        final File file = new File(mContext.getFilesDir(), FILENAME);
        final String json = toJsonString(shapeList);
        saveString(file, json);
    }

    @WorkerThread
    @NonNull
    private List<Shape> internalRestoreShapes() {
        final File file = new File(mContext.getFilesDir(), FILENAME);
        if (file.exists() && file.canRead()) {
            final String json = readString(file);
            return fromJsonString(json);
        } else {
            return Collections.emptyList();
        }
    }

    @NonNull
    private String toJsonString(@NonNull List<Shape> shapeList) {
        final JSONArray result;
        try {
            result = new JSONArray();
            for (Shape shape : shapeList) {
                final JSONObject jsonObject = new JSONObject();
                jsonObject.put(SHAPE_X, shape.getX());
                jsonObject.put(SHAPE_Y, shape.getY());
                jsonObject.put(SHAPE_COLOR, shape.getColor());

                if (shape instanceof Circle) {
                    jsonObject.put(SHAPE_TYPE, TYPE_CIRCLE);
                    final Circle circle = (Circle) shape;
                    jsonObject.put(CIRCLE_RADIUS, circle.getRadius());
                } else if (shape instanceof Rectangle) {
                    jsonObject.put(SHAPE_TYPE, TYPE_RECTANGLE);
                    final Rectangle rectangle = (Rectangle) shape;
                    jsonObject.put(RECT_WIDTH, rectangle.getWidth());
                    jsonObject.put(RECT_HEIGHT, rectangle.getHeight());
                } else if (shape instanceof Text) {
                    jsonObject.put(SHAPE_TYPE, TYPE_TEXT);
                    final Text text = (Text) shape;
                    jsonObject.put(TEXT_TEXT, text.getText());
                    jsonObject.put(TEXT_FONT_SIZE, text.getFontSize());
                }

                result.put(jsonObject);
            }
            return result.toString();
        } catch (JSONException e) {
            return "[]";
        }
    }

    @NonNull
    private List<Shape> fromJsonString(@NonNull String jsonArrayString) {
        final List<Shape> shapeList = new ArrayList<>();

        try {
            final JSONArray jsonArray = new JSONArray(jsonArrayString);
            final int shapeCount = jsonArray.length();
            for (int i = 0; i < shapeCount; i++) {
                final JSONObject shape = jsonArray.getJSONObject(i);
                final float x = (float) shape.getDouble(SHAPE_X);
                final float y = (float) shape.getDouble(SHAPE_Y);
                final int color = shape.getInt(SHAPE_COLOR);
                final String type = shape.getString(SHAPE_TYPE);

                switch (type) {
                    case TYPE_CIRCLE:
                        final float radius = (float) shape.getDouble(CIRCLE_RADIUS);
                        shapeList.add(new Circle(x, y, color, radius));
                        break;

                    case TYPE_RECTANGLE:
                        final float width = (float) shape.getDouble(RECT_WIDTH);
                        final float height = (float) shape.getDouble(RECT_HEIGHT);
                        shapeList.add(new Rectangle(x, y, color, width, height));
                        break;

                    case TYPE_TEXT:
                        final String text = shape.getString(TEXT_TEXT);
                        final float fontSize = (float) shape.getDouble(TEXT_FONT_SIZE);
                        shapeList.add(new Text(x, y, color, fontSize, text));
                        break;
                }
            }
        } catch (JSONException e) {
            Log.e(TAG, "Failed to serialize shape list", e);
        }

        return shapeList;
    }

    @WorkerThread
    private void saveString(@NonNull File file, @NonNull String string) {
        try (FileOutputStream out = new FileOutputStream(file)) {
            out.write(string.getBytes());
        } catch (IOException e) {
            Log.e(TAG, "Failed to save file", e);
        }
    }

    @WorkerThread
    @NonNull
    private String readString(@NonNull File file) {
        try (FileInputStream in = new FileInputStream(file)) {
            final BufferedReader reader = new BufferedReader(new InputStreamReader(in));
            final StringBuilder builder = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                builder.append(line);
            }

            return builder.toString();
        } catch (IOException e) {
            Log.e(TAG, "Failed to read file", e);
            return "[]";
        }
    }

    public interface OnSaveFinishedCallback {
        void onSaveFinished();
    }

    public interface OnRestoreFinishedCallback {
        void onRestoreFinished(@NonNull List<Shape> shapeList);
    }
}

File: app/src/main/java/com/codepath/shapes/io/ShapeSerializer.kt
package com.codepath.shapes.io

import android.content.Context
import android.os.AsyncTask
import android.support.annotation.MainThread
import android.support.annotation.WorkerThread
import android.util.Log
import com.codepath.shapes.shape.Circle
import com.codepath.shapes.shape.Rectangle
import com.codepath.shapes.shape.Shape
import com.codepath.shapes.shape.Text
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.File
import java.io.IOException

typealias ShapeList = List<Shape>
typealias OnSaveShapesCallback = () -> Unit
typealias OnRestoreShapesCallBack = (ShapeList) -> Unit

class ShapeSerializer(private val mContext: Context) {
    private val TAG = "ShapeSerializer"

    private val SHAPE_X = "x"
    private val SHAPE_Y = "y"
    private val SHAPE_COLOR = "color"
    private val SHAPE_TYPE = "type"
    private val CIRCLE_RADIUS = "radius"
    private val RECT_WIDTH = "width"
    private val RECT_HEIGHT = "height"
    private val TEXT_TEXT = "text"
    private val TEXT_FONT_SIZE = "fontSize"
    private val TYPE_CIRCLE = "circle"
    private val TYPE_RECTANGLE = "rectangle"
    private val TYPE_TEXT = "text"

    private val FILENAME = "shapes.json"

    @MainThread
    fun saveShapes(shapeList: ShapeList, callback: OnSaveShapesCallback) {
        object : AsyncTask<ShapeList, Nothing, Unit>() {
            @SafeVarargs
            override fun doInBackground(vararg params: ShapeList) = internalSaveShapes(params.first())

            override fun onPostExecute(result: Unit?) = callback()
        }.execute(shapeList)
    }

    @MainThread
    fun restoreShapes(callback: OnRestoreShapesCallBack) {
        object : AsyncTask<Void, Nothing, ShapeList>() {
            override fun doInBackground(vararg params: Void) = internalRestoreShapes()
            override fun onPostExecute(shapeList: ShapeList) = callback(shapeList)
        }.execute()
    }

    @WorkerThread
    private fun internalSaveShapes(shapeList: ShapeList) = File(mContext.filesDir, FILENAME).writeText(shapeList.toJsonString())

    @WorkerThread
    private fun internalRestoreShapes(): ShapeList {
        val file = File(mContext.filesDir, FILENAME)
        return if (file.exists() && file.canRead()) {
            val json = readString(file)
            fromJsonString(json)
        } else {
            emptyList()
        }
    }

    private fun ShapeList.toJsonString(): String = try {
        JSONArray().apply {
            map { shape -> shape.toJsonObject() }.forEach { put(it) }
        }.toString()
    } catch (e: JSONException) {
        "[]"
    }

    private fun Shape.toJsonObject() = JSONObject().apply {
        put(SHAPE_X to x,
                SHAPE_Y to y,
                SHAPE_COLOR to color)

        when (this@toJsonObject) {
            is Circle -> {
                put(SHAPE_TYPE to TYPE_CIRCLE,
                        CIRCLE_RADIUS to radius)
            }

            is Rectangle -> {
                put(SHAPE_TYPE to TYPE_RECTANGLE,
                        RECT_WIDTH to width,
                        RECT_HEIGHT to height)
            }

            is Text -> {
                put(SHAPE_TYPE to TYPE_TEXT,
                        TEXT_TEXT to text,
                        TEXT_FONT_SIZE to fontSize)
            }
        }
    }

    private fun fromJsonString(jsonArrayString: String): ShapeList {
        val shapeList = mutableListOf<Shape>()

        try {
            val jsonArray = JSONArray(jsonArrayString)
            for (i in 0 until jsonArray.length()) {
                with(jsonArray.getJSONObject(i)) {
                    val x = getFloat(SHAPE_X)
                    val y = getFloat(SHAPE_Y)
                    val color = getInt(SHAPE_COLOR)
                    val type = getString(SHAPE_TYPE)

                    val shape = when (type) {
                        TYPE_CIRCLE -> {
                            val radius = getFloat(CIRCLE_RADIUS)
                            Circle(x, y, color, radius)
                        }

                        TYPE_RECTANGLE -> {
                            val width = getFloat(RECT_WIDTH)
                            val height = getFloat(RECT_HEIGHT)
                            Rectangle(x, y, color, width, height)
                        }

                        TYPE_TEXT -> {
                            val text = getString(TEXT_TEXT)
                            val fontSize = getFloat(TEXT_FONT_SIZE)
                            Text(x, y, color, fontSize, text)
                        }

                        else -> null
                    }
                    shape?.let { shapeList += it }
                }
            }
        } catch (e: JSONException) {
            Log.e(TAG, "Failed to serialize shape list", e)
        }

        return shapeList
    }

    @WorkerThread
    private fun readString(file: File) = try {
        file.readText()
    } catch (e: IOException) {
        Log.e(TAG, "Failed to read file", e)
        "[]"
    }

    private fun JSONObject.getFloat(name: String) = getDouble(name).toFloat()
    private fun JSONObject.put(vararg values: Pair<String, Any>) {
        values.forEach { put(it.first, it.second) }
    }
}
